### PR TITLE
use ForemanApi from apypie

### PR DESCRIPTION
**Pull request by evgeni**
includes #1797

### CODE REVIEW

File: plugins/module_utils/_apypie.py [hunk 1]
```diff
@@ -921,3 +921,188 @@ def path_with_params(self, params=None):
...
+
+        :param resource: Plural name of the api resource to show
+        :type resource: str
+        :param search: Search string as accepted by the API to limit the results
+        :type search: str, optional
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+        :type params: Union[dict,None], optional
+        """
+        PER_PAGE = 2 << 31
+        payload = {'per_page': PER_PAGE}
```
Comment (ekohl) [inline plugins/module_utils/_apypie.py]: Foreman supports `per_page=all`, but Katello has its own mechanism using `full_result=true`. I started https://github.com/Katello/katello/pull/11126 to unify this, but I don't think that's ready to be merged any time soon.

File: plugins/module_utils/foreman_helper.py [hunk 2]
```diff
@@ -700,18 +677,7 @@ def list_resource(self, resource, search=None, params=None):
...
         """
 
-        if params is None:
-            params = {}
-        else:
-            params = params.copy()
-
-        if search is not None:
-            params['search'] = search
-        params['per_page'] = PER_PAGE
```
Comment (ekohl) [inline plugins/module_utils/foreman_helper.py:710]: The constant definition is now unused.

File: plugins/module_utils/_apypie.py [hunk 3]
```diff
@@ -921,3 +921,188 @@ def path_with_params(self, params=None):
...
+    def show(self, resource, resource_id, params=None):
+        """
+        Execute the ``show`` action on an entity.
+
+        :param resource: Plural name of the api resource to show
+        :type resource: str
+        :param resource_id: The ID of the entity to show
+        :type resource_id: int
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+        :type params: Union[dict,None], optional
```
Comment (ekohl) [inline plugins/module_utils/_apypie.py]: There's also `Optional[dict]` according to https://docs.python.org/3/library/typing.html#typing.Optional so is this also sufficient?
```suggestion
        :type params: Optional[dict]
```

File: plugins/module_utils/foreman_helper.py [hunk 4]
```diff
@@ -1024,7 +990,7 @@ def _validate_supported_payload(self, resource, action, payload):
         :return: The payload as it can be submitted to the API
         :rtype: dict
         """
-        filtered_payload = self._resource_prepare_params(resource, action, payload)
+        filtered_payload = self.foremanapi._resource_prepare_params(resource, action, payload)
```
Comment (ekohl) [inline plugins/module_utils/foreman_helper.py]: Calling a private API isn't very nice.

**Review by ekohl — COMMENTED** [review]
Consider this some of the notes I made while reading this. I realize it's very much a work in progress and should result in a real PR to apypie itself before updating the vendored copy.

File: plugins/module_utils/foreman_helper.py (hunk 2 above)
Comment (evgeni) [reply plugins/module_utils/foreman_helper.py:710]: That's not correct ;)
```
plugins/modules/activation_key.py:from ansible_collections.theforeman.foreman.plugins.module_utils.foreman_helper import KatelloEntityAnsibleModule, PER_PAGE
```

File: plugins/module_utils/_apypie.py (hunk 3 above)
Comment (evgeni) [reply plugins/module_utils/_apypie.py]: Yeah, that's the same.

File: plugins/module_utils/foreman_helper.py (hunk 4 above)
Comment (evgeni) [reply plugins/module_utils/foreman_helper.py]: so let's not do that! :)
(fixed)

**Review by ekohl — APPROVED** [review]

**Pull request merged**